genqlient crash in preprocessQueryDocument

Stacktrace:
```
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x40 pc=0x6d064d]

goroutine 1 [running]:
github.com/Khan/genqlient/generate.(*generator).preprocessQueryDocument.func1(0xc000174240, 0xc0000dac00)
	/home/benkraft/.go/pkg/mod/github.com/!khan/genqlient@v0.4.0/generate/generate.go:188 +0x2d
[...]
```
which is here:
```go
		fieldType := g.schema.Types[field.Definition.Type.Name()]
```
I assume some particular case gets us here before validation, or validation doesn't catch something it should.  Anyway, we should give a proper error that the type is undefined.

--- a/generate/generate.go
+++ b/generate/generate.go
@@ -224,16 +224,40 @@ func (g *generator) preprocessQueryDocument(doc *ast.QueryDocument) {
 	validator.Walk(g.schema, doc, &observers)
 }
 
+// validateOperation checks for a few classes of operations that gqlparser
+// considers valid but we don't allow, and returns an error if this operation
+// is invalid for genqlient's purposes.
+func (g *generator) validateOperation(op *ast.OperationDefinition) error {
+	opType, err := g.baseTypeForOperation(op.Operation)
+	switch {
+	case err != nil:
+		// (e.g. operation has subscriptions, which we don't support)
+		return err
+	case opType == nil:
+		// gqlparser should err here, but doesn't [1], so we err to prevent
+		// panics later.
+		// TODO(benkraft): Remove once gqlparser is fixed.
+		// [1] https://github.com/vektah/gqlparser/issues/221
+		return errorf(op.Position, "schema has no %v type", op.Operation)
+	}
+
+	if op.Name == "" {
+		return errorf(op.Position, "operations must have operation-names")
+	} else if goKeywords[op.Name] {
+		return errorf(op.Position, "operation name must not be a go keyword")
+	}
+
+	return nil
+}
+
 // addOperation adds to g.Operations the information needed to generate a
 // genqlient entrypoint function for the given operation.  It also adds to
 // g.typeMap any types referenced by the operation, except for types belonging
 // to named fragments, which are added separately by Generate via
 // convertFragment.
 func (g *generator) addOperation(op *ast.OperationDefinition) error {
-	if op.Name == "" {
-		return errorf(op.Position, "operations must have operation-names")
-	} else if goKeywords[op.Name] {
-		return errorf(op.Position, "operation name must not be a go keyword")
+	if err := g.validateOperation(op); err != nil {
+		return err
 	}
 
 	queryDoc := &ast.QueryDocument{

--- a/generate/testdata/errors/NoMutationType.graphql
+++ b/generate/testdata/errors/NoMutationType.graphql
@@ -1,0 +1,1 @@
+mutation M { g }

--- a/generate/testdata/errors/NoMutationType.schema.graphql
+++ b/generate/testdata/errors/NoMutationType.schema.graphql
@@ -1,0 +1,1 @@
+type Query { f: String }

--- a/generate/testdata/errors/NoQueryType.graphql
+++ b/generate/testdata/errors/NoQueryType.graphql
@@ -1,0 +1,1 @@
+query Q { f }

--- a/generate/testdata/errors/NoQueryType.schema.graphql
+++ b/generate/testdata/errors/NoQueryType.schema.graphql
@@ -1,0 +1,1 @@
+type T { f: String }

--- a/generate/testdata/snapshots/TestGenerateErrors-NoMutationType-graphql
+++ b/generate/testdata/snapshots/TestGenerateErrors-NoMutationType-graphql
@@ -1,0 +1,1 @@
+testdata/errors/NoMutationType.graphql:1: schema has no mutation type

--- a/generate/testdata/snapshots/TestGenerateErrors-NoQueryType-graphql
+++ b/generate/testdata/snapshots/TestGenerateErrors-NoQueryType-graphql
@@ -1,0 +1,1 @@
+testdata/errors/NoQueryType.graphql:1: schema has no query type

